src/mixpanel-core.js

```diff
@@ -1106,6 +1108,362 @@ MixpanelLib.prototype.track = addOptOutCheckMixpanelLib(function(event_name, pro
+     *
+     * Each event automatically tracks:
+     * - $duration: Seconds from first to last heartbeat call
+     * - $heartbeats: Number of heartbeat calls made
+     * - $contentId: The contentId parameter
+     *
+     * @function heartbeat
+     * @memberof mixpanel
+     * @param {String} eventName The name of the event to track
+     * @param {String} contentId Unique identifier for the content being tracked
+     * @param {Object} [props] Properties to aggregate with existing data
+     * @param {Object} [options] Configuration options
+     * @param {Number} [options.timeout] Timeout in milliseconds (default 30000)
+     * @param {Boolean} [options.forceFlush] Force immediate flush after aggregation
+     * @returns {Void}
+     *
+     * @example
+     * // Basic video tracking
+     * mixpanel.heartbeat('video_watch', 'video_123', { quality: 'HD' });
+     * mixpanel.heartbeat('video_watch', 'video_123', { position: 30 });
+     * // After 30 seconds: { quality: 'HD', position: 30, $duration: 30, $heartbeats: 2 }
+     *
+     * @example
+     * // Force immediate flush
+     * mixpanel.heartbeat('podcast_listen', 'episode_123', { platform: 'mobile' }, { forceFlush: true });
+     *
+     * @example
+     * // Custom timeout (60 seconds)
+     * mixpanel.heartbeat('video_watch', 'video_123', { quality: 'HD' }, { timeout: 60000 });
+     */
+    this.heartbeat = function(eventName, contentId, props, options) {
+        return self._heartbeat_impl(eventName, contentId, props, options);
+    };
+
+};
+
+/**
+ * Sets up page unload handlers for heartbeat auto-flush
+ * @private
+ */
+MixpanelLib.prototype._setup_heartbeat_unload_handlers = function() {
+    if (this._heartbeat_unload_setup) {
+        return;
+    }
+    this._heartbeat_unload_setup = true;
+
+    var self = this;
+    var handleUnload = function() {
+        self._heartbeat_log('Page unload detected, flushing all heartbeat events');
+        self._heartbeat_flush_all('pageUnload', true);
+    };
+
+    // Multiple event handlers for cross-browser compatibility
+    if (window.addEventListener) {
+        window.addEventListener('beforeunload', handleUnload);
+        window.addEventListener('pagehide', handleUnload);
+        window.addEventListener('visibilitychange', function() {
+            if (document.visibilityState === 'hidden') {
+                handleUnload();
+            }
+        });
+    }
+};
+
+/**
+ * Gets heartbeat event storage from persistence
+ * @private
+ */
+MixpanelLib.prototype._heartbeat_get_storage = function() {
+    var stored = this['persistence'].props[HEARTBEAT_QUEUE_KEY];
+    return stored && typeof stored === 'object' ? stored : {};
+};
+
+/**
+ * Saves heartbeat events to persistence
+ * @private
+ */
+MixpanelLib.prototype._heartbeat_save_storage = function(data) {
+    var current_props = {};
+    current_props[HEARTBEAT_QUEUE_KEY] = data;
+    this['persistence'].register(current_props);
```

Comment:
I don't know that persisting heartbeat data in the cookie is a great move here. If this system really needs browser persistence, we can probably hook into the existing batch queue utils with some modification, which support indexeddb and localstorage. That might be another overall approach too: make heartbeats a special queue in the request batcher, an event which doesn't get flushed on a regular schedule, and every time a new heartbeat happens, the queued event is rewritten. @jakewski any thoughts?